typing/typetexp.ml

```diff
@@ -267,14 +270,27 @@ end = struct
       associate row_context p;
       p.univar
     with Not_found ->
-      instance (fst (TyVarMap.find name !used_variables))
+      let (v, _, unused) = TyVarMap.find name !used_variables in
+      unused := false;
+      instance v
       (* This call to instance might be redundant; all variables
          inserted into [used_variables] are non-generic, but some
          might get generalized. *)
 
-  let remember_used name v loc =
+  let remember_used ?check name v loc =
     assert (not_generic v);
-    used_variables := TyVarMap.add name (v, loc) !used_variables
+    let unused = match check with
+      | None -> ref false
+      | Some check_loc ->
+        let unused = ref true in
+        !Env.add_delayed_check_forward begin fun () ->
+            let warn = Warnings.Unused_type_declaration ("'" ^ name) in
```

Comment:
This works because `add_delayed_check_forward` stores the warning state, but I would propose to check first if the warning is active with `Warnings.is_active (Warnings.Unused_type_declaration "")`, and only registers the check it may trigger.